### ISSUE TITLE
[dv] Bias in randomization of MuBi values

### ISSUE THREAD

**Issue by vogelpi** [lowRISC/opentitan#16565]
During the V2/V2S work for CSRNG, it occurred to @andreaskurth , @marnovandermaas and me that there is bias when randomizing MuBi signals in DV (see also https://github.com/lowRISC/opentitan/pull/16531#discussion_r1029937898). For example, when using
```
item.my_mubi = get_rand_mubi4_val(.t_weight(4), .f_weight(4), .other_weight(0));
```
or
```
`DV_CHECK_MEMBER_RANDOMIZE_FATAL(my_mubi, item)
```
or
```
`DV_CHECK_MEMBER_RANDOMIZE_WITH_FATAL(my_mubi, my_mubi inside {MuBi4True, MuBi4False};, item)
```
we seemed to get ratios of  `MuBi4True` / `MuBi4False` of ~1/15 or ~1/20 wit Xcelium.

Due to a lack of time we didn't further investigate. But I was wondering is this something other people have experienced as well? Or did we just miss something? It could also be something Xcelium specific.

Maybe @sriyerg , @weicaiyang or @matutem might know something.

**Comment by weicaiyang** [comment]
We got a very well distribute result when I tried below with Xcelium. The ratio is almost 50/50
```
repeat (100) $display("mubi %0d", get_rand_mubi4_val(.t_weight(4), .f_weight(4), .other_weight(0)));
$finish;
```
or

```
 prim_mubi_pkg::mubi4_t my_mubi;
task body();
repeat (100)begin
 `DV_CHECK_MEMBER_RANDOMIZE_WITH_FATAL(my_mubi, my_mubi inside {prim_mubi_pkg::MuBi4True, 
   prim_mubi_pkg::MuBi4False};)
  $display("mubi again %0d", my_mubi);
end
$finish;
endtask : body

```


-------
If we don't have addition constraint for this one, the chance to have true/false is very low, since all values are distributed equally.
```
`DV_CHECK_MEMBER_RANDOMIZE_FATAL(my_mubi, item)
```

**Comment by matutem** [comment]
I agree with @weicaiyang: in the three examples shown in the issue description, the first and third should have equal distribution of True and False, while the second should have a uniform distribution for all 16 values.

I would be inclined to close this unless we have more specific proof of xcelium misbehavior.

**Comment by andreaskurth** [comment]
Triaged for `csrng`. Assigning to M2.5 with https://github.com/lowRISC/opentitan/labels/Priority%3AP4 because we should keep this in mind while working on M2.5, but if we don't *have* to solve it for M2.5, I think we shouldn't.

**Comment by msfschaffner** [comment]
I don't think this is critical for M2.5 - postponing.

**Comment by vogelpi** [comment]
I am pretty sure this got addressed by @rswarbrick in this PR here #29330 which corrected the weights and fixed Xcelium warnings.

Note this PR actually touched the `_DV_MUBI_DIST` macro in `cip_macros.svh` but these get called by `DV_MUBIWIDTH_DIST` and `get_rand_mubiWIDTH_val()` defined in `cip_base_pkg.sv`. This last function is the one for which the issue was reported in the first message in this issue.

I am thus closing this issue as completed.
